src/graphics/opengl/RendererGL.cpp

```diff
@@ -1202,17 +1074,50 @@ namespace Graphics {
 		return new OGL::InstanceBuffer(size, usage);
 	}
 
-	// XXX very heavy. in the future when all GL calls are made through the
-	// renderer, we can probably do better by trackingn current state and
-	// only restoring the things that have changed
-	void RendererOGL::PushState()
+	UniformBuffer *RendererOGL::CreateUniformBuffer(Uint32 size, BufferUsage usage)
+	{
+		m_stats.AddToStatCount(Stats::STAT_CREATE_BUFFER, 1);
+		return new OGL::UniformBuffer(size, usage);
+	}
+
+	MeshObject *RendererOGL::CreateMeshObject(VertexBuffer *v, IndexBuffer *i)
+	{
+		m_stats.AddToStatCount(Stats::STAT_CREATE_BUFFER, 1);
+		return new OGL::MeshObject(v, i);
+	}
+
+	MeshObject *RendererOGL::CreateMeshObjectFromArray(const VertexArray *vertexArray, IndexBuffer *indexBuffer, BufferUsage usage)
+	{
+		VertexBufferDesc desc = VertexBufferDesc::FromAttribSet(vertexArray->GetAttributeSet());
+		desc.numVertices = vertexArray->GetNumVerts();
+		desc.usage = usage;
+
+		Graphics::VertexBuffer *vertexBuffer = CreateVertexBuffer(desc);
+		vertexBuffer->Populate(*vertexArray);
+
+		return CreateMeshObject(vertexBuffer, indexBuffer);
+	}
+
+	OGL::UniformBuffer *RendererOGL::GetLightUniformBuffer()
+	{
+		return m_lightUniformBuffer.Get();
+	}
+
+	OGL::UniformLinearBuffer *RendererOGL::GetDrawUniformBuffer(Uint32 size)
 	{
-		// empty since viewport handling is now external, evaluate if renderer will need to save any custom state
+		for (auto &buffer : m_drawUniformBuffers)
+			if (buffer->FreeSize() >= size)
+				return buffer.get();
+
+		auto *buffer = new OGL::UniformLinearBuffer(1 << 20);
+		buffer->IncRefCount();
```

Comment:
Why not do
`std::vector<RefCountedPtr<OGL::UniformLinearBuffer>> m_drawUniformBuffers` ?
It seems that the entire overhead (I mean `mutable std::atomic<int> m_refCount`) is still with us.
But at least you don't have to manually increase the smart pointer germ counter through the raw pointer.

Reply:
Good catch with the update; I didn't get a chance to comment earlier, but this is in fact a left-over "optimization" from an earlier version when Materials stored `RefCountedPtr<UniformBuffer>` instead of a raw pointer.
Ownership of the linear UBO should stay entirely with the Renderer (hence the API using raw pointers), but the ref count increment was to prevent it from being accidentally deleted by Material constructing a RefCountedPtr from a raw pointer, as is the general paradigm with objects returned from the Renderer interface.
This IncRefCount() line can *most likely* be deleted without impacting the code at all, but I'd have to swap over to the branch and trace it just to be sure.